File: DHBCollision.js
class DHBCollision extends Collision {
    //Disc-HorizontalBorder collision
    constructor(body1, body2) {
        if (!(body1 instanceof Disc && body2 instanceof HorizontalBorder)) {
            throw `Wrong collision type: Not a Disk-HB collision`;
        } else {
            super(body1, body2);
            if (body2.extendsTo == Way.up) {
                this.collisionNormal = new Vector(0, 1);
            } else if (body2.extendsTo == Way.down) {
                this.collisionNormal = new Vector(0, -1);
            } else {
                this.collisionNormal = body1.center.y > body2.center.y ? new Vector(0, 1) : new Vector(0, -1);
            }

            this.relativeVelocity = Vector.sub(this.body1.velocity, this.body2.velocity);
            this.velocityAlongNormal = Vector.dot(this.collisionNormal, this.relativeVelocity);
            this.penetrationDepth = this.body1.radius - Math.abs(this.body1.center.y - this.body2.center.y);
        }
    }
}

File: pyrl/DHBCollision.py
from Collision import Collision
from HorizontalBorder import HorizontalBorder
from Vector import Vector
from Way import Way
from Disc import Disc

class DHBCollision(Collision):
    #Disc-VerticalBorder collision
    def __init__(self, body1, body2):
        assert isinstance(body1, Disc) and isinstance(body2, HorizontalBorder)
        super.__init__(body1, body2)
        if body2.extends_to == Way.left:
            self.collision_normal = Vector(1, 0)
        elif body2.extends_to == Way.right:
            self.collision_normal = Vector(-1, 0);
        else:
            self.collision_normal = Vector(1, 0) if body1.center.x > body2.center.x else Vector(-1, 0)
        
        self.relative_velocity = Vector.sub(self.body1.velocity, self.body2.velocity)
        self.velocity_along_normal = Vector.dot(self.collision_normal, self.relative_velocity)
        self.penetration_depth = self.body1.radius - abs(self.body1.center.x - self.body2.center.x)
    
    

# -*- coding: utf-8 -*-

